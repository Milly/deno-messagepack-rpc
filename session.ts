import { Reservator } from "https://deno.land/x/reservator@v0.1.0/mod.ts";
import {
  DecodeStream,
  encode,
} from "https://deno.land/x/messagepack@v0.1.0/mod.ts";
import {
  buildResponseMessage,
  isNotificationMessage,
  isRequestMessage,
  isResponseMessage,
  Message,
  MessageError,
  MessageId,
  MessageResult,
  NotificationMessage,
  RequestMessage,
  ResponseMessage,
} from "./message.ts";
import { Dispatcher } from "./dispatcher.ts";

export type SessionOptions = {
  onUnexpectedError?: (message: Message, err: Error) => void | Promise<void>;
  onUnexpectedMessage?: (message: unknown) => void | Promise<void>;
};

export class Session {
  #reservator: Reservator<MessageId, unknown>;
  #rstream: ReadableStream<Uint8Array>;
  #wstream: WritableStream<Uint8Array>;
  #onUnexpectedError: (message: Message, err: Error) => void | Promise<void>;
  #onUnexpectedMessage: (message: unknown) => void | Promise<void>;

  dispatcher: Dispatcher;

  constructor(
    rstream: ReadableStream<Uint8Array>,
    wstream: WritableStream<Uint8Array>,
    options: SessionOptions = {},
  ) {
    const {
      onUnexpectedError = defaultOnUnexpectedError,
      onUnexpectedMessage = () => {},
    } = options;
    this.#reservator = new Reservator();
    this.#rstream = rstream;
    this.#wstream = wstream;
    this.#onUnexpectedError = onUnexpectedError;
    this.#onUnexpectedMessage = onUnexpectedMessage;
    this.dispatcher = {};
  }

  async #dispatch(
    method: string,
    params: unknown[],
  ): Promise<[MessageResult, MessageError]> {
    try {
      return [
        await this.dispatcher[method](...params),
        null,
      ];
    } catch (err: unknown) {
      if (err instanceof TypeError && !hasMethod(this.dispatcher, method)) {
        return [
          null,
          new Error(
            `No MessagePack-RPC method '${method}' exists`,
          ),
        ];
      }
      return [null, err];
    }
  }

  async send(message: Message): Promise<void> {
    const data = encode(message);
    const writer = this.#wstream.getWriter();
    await writer.ready;
    await writer.write(data);
    writer.releaseLock();
  }

  wait(msgid: number): Promise<unknown> {
    return this.#reservator.reserve(msgid);
  }

  #handleRequestMessage(message: RequestMessage): void {
    (async () => {
      const [_, msgid, method, params] = message;
      const [result, error] = await this.#dispatch(method, params);
      const response = buildResponseMessage(msgid, error, result);
      await this.send(response);
    })().catch((err) => this.#onUnexpectedError(message, err));
  }

  #handleNotificationMessage(
    message: NotificationMessage,
  ): void {
    (async () => {
      const [_, method, params] = message;
      const [__, err] = await this.#dispatch(method, params);
      if (err) {
        if (err instanceof Error) {
          throw err;
        } else {
          throw new Error(`${err}`);
        }
      }
    })().catch((err) => this.#onUnexpectedError(message, err));
  }

  #handleResponseMessage(message: ResponseMessage): void {
    const [_, msgid, err, result] = message;
    if (err) {
      let error: Error;
      if (err instanceof Error) {
        error = err;
      } else if (Array.isArray(err) && err.length === 2) {
        error = new Error(`${err[1]}`);
        error.name = `${err[0]}`;
      } else {
        error = new Error(`${err}`);
      }
      this.#reservator.reject(msgid, error);
    } else {
      this.#reservator.resolve(msgid, result);
    }
  }

  start(options: { signal?: AbortSignal } = {}): Promise<void> {
    const { signal } = options;
    const sink = new WritableStream({
      write: (message) => {
        if (isRequestMessage(message)) {
          this.#handleRequestMessage(message);
        } else if (isNotificationMessage(message)) {
          this.#handleNotificationMessage(message);
        } else if (isResponseMessage(message)) {
          this.#handleResponseMessage(message);
        } else {
          // Unknown message. Ignore it for forward compatibility.
          this.#onUnexpectedMessage(message);
        }
      },
    });
    return this.#rstream.pipeThrough(new DecodeStream()).pipeTo(sink, {
      signal,
    });
  }
}

function hasMethod(obj: unknown, method: string): boolean {
  return Object.prototype.hasOwnProperty.call(obj, method);
}

function defaultOnUnexpectedError(message: Message, err: Error): void {
  console.error(
    `Unexpected error for MessagePack-RPC message ${message}: ${err}`,
  );
}
