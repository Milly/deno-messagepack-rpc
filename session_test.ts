import {
  assert,
  assertEquals,
  assertRejects,
} from "https://deno.land/std@0.185.0/testing/asserts.ts";
import {
  assertSnapshot,
} from "https://deno.land/std@0.185.0/testing/snapshot.ts";
import { assertNumber } from "https://deno.land/x/unknownutil@v2.1.0/assert.ts";
import { deferred, delay } from "https://deno.land/std@0.185.0/async/mod.ts";
import { decode, encode } from "https://deno.land/x/messagepack@v0.1.0/mod.ts";
import {
  buildNotificationMessage,
  buildRequestMessage,
  buildResponseMessage,
} from "./message.ts";
import { Session } from "./session.ts";

Deno.test("Session", async (t) => {
  await t.step(
    "invokes a corresponding method and sends back a result when a request is received",
    async (t) => {
      const output: unknown[] = [];
      const session = new Session(
        new ReadableStream({
          start(controller) {
            controller.enqueue(encode(buildRequestMessage(0, "sum", [1, 2])));
            controller.close();
          },
        }),
        new WritableStream({
          write(chunk) {
            output.push(decode(chunk));
          },
        }),
      );

      let called = false;
      session.dispatcher = {
        sum(x, y) {
          called = true;
          assertNumber(x);
          assertNumber(y);
          return x + y;
        },
      };

      await session.start();

      assert(called);
      assertSnapshot(t, output);
    },
  );

  await t.step(
    "invokes a corresponding method and sends back an error result when a request is received (error)",
    async (t) => {
      const output: unknown[] = [];
      const session = new Session(
        new ReadableStream({
          start(controller) {
            controller.enqueue(encode(buildRequestMessage(0, "sum", [1, 2])));
            controller.close();
          },
        }),
        new WritableStream({
          write(chunk) {
            output.push(decode(chunk));
          },
        }),
      );

      let called = false;
      session.dispatcher = {
        sum() {
          called = true;
          throw new Error("Failed");
        },
      };

      await session.start();

      assert(called);
      assertSnapshot(t, output);
    },
  );

  await t.step(
    "sends back an error response when a request is received but no corresponding method exists",
    async (t) => {
      const output: unknown[] = [];
      const session = new Session(
        new ReadableStream({
          start(controller) {
            controller.enqueue(encode(buildRequestMessage(0, "sum", [1, 2])));
            controller.close();
          },
        }),
        new WritableStream({
          write(chunk) {
            output.push(decode(chunk));
          },
        }),
      );

      await session.start();

      assertSnapshot(t, output);
    },
  );

  await t.step(
    "invokes a corresponding method when a notification is received",
    async () => {
      const output: unknown[] = [];
      const session = new Session(
        new ReadableStream({
          start(controller) {
            controller.enqueue(encode(buildNotificationMessage("sum", [1, 2])));
            controller.close();
          },
        }),
        new WritableStream({
          write(chunk) {
            output.push(decode(chunk));
          },
        }),
      );

      let called = false;
      session.dispatcher = {
        sum(x, y) {
          called = true;
          assertNumber(x);
          assertNumber(y);
          return x + y;
        },
      };

      await session.start();

      assert(called);
      assertEquals(output, []);
    },
  );

  await t.step(
    "invokes a corresponding method and invokes 'onUnexpectedError()' when a notification is received (error)",
    async () => {
      let onUnexpectedErrorCalled = false;
      const output: unknown[] = [];
      const session = new Session(
        new ReadableStream({
          start(controller) {
            controller.enqueue(encode(buildNotificationMessage("sum", [1, 2])));
            controller.close();
          },
        }),
        new WritableStream({
          write(chunk) {
            output.push(decode(chunk));
          },
        }),
        {
          onUnexpectedError() {
            onUnexpectedErrorCalled = true;
          },
        },
      );

      let called = false;
      session.dispatcher = {
        sum() {
          called = true;
          throw new Error("Failed");
        },
      };

      await session.start();

      assert(called);
      assert(onUnexpectedErrorCalled);
      assertEquals(output, []);
    },
  );

  await t.step(
    "invokes 'onUnexpectedError()' when a notification is received but no corresponding method exists",
    async () => {
      let onUnexpectedErrorCalled = false;
      const output: unknown[] = [];
      const session = new Session(
        new ReadableStream({
          start(controller) {
            controller.enqueue(encode(buildNotificationMessage("sum", [1, 2])));
            controller.close();
          },
        }),
        new WritableStream({
          write(chunk) {
            output.push(decode(chunk));
          },
        }),
        {
          onUnexpectedError() {
            onUnexpectedErrorCalled = true;
          },
        },
      );

      await session.start();

      assert(onUnexpectedErrorCalled);
      assertEquals(output, []);
    },
  );

  await t.step(
    "invokes onUnexpectedMessage and ignore when unknown message is received",
    async (t) => {
      let onUnexpectedMessageCalled = false;
      const output: unknown[] = [];
      const session = new Session(
        new ReadableStream({
          start(controller) {
            controller.enqueue(new Uint8Array([0, 1, 2, 3, 4]));
            controller.enqueue(encode(buildRequestMessage(0, "sum", [1, 2])));
            controller.enqueue(new Uint8Array([5, 6, 7, 8, 9]));
            controller.close();
          },
        }),
        new WritableStream({
          write(chunk) {
            output.push(decode(chunk));
          },
        }),
        {
          onUnexpectedMessage() {
            onUnexpectedMessageCalled = true;
          },
        },
      );

      let called = false;
      session.dispatcher = {
        sum(x, y) {
          called = true;
          assertNumber(x);
          assertNumber(y);
          return x + y;
        },
      };

      await session.start();

      assert(called);
      assert(onUnexpectedMessageCalled);
      assertSnapshot(t, output);
    },
  );

  await t.step(
    "resolves a promise returned by 'wait()' method when a response is received",
    async () => {
      const guard = deferred<void>();
      const output: unknown[] = [];
      const session = new Session(
        new ReadableStream({
          async start(controller) {
            await guard;
            controller.enqueue(
              encode(buildResponseMessage(0, null, "Success")),
            );
            controller.close();
          },
        }),
        new WritableStream({
          write(chunk) {
            output.push(decode(chunk));
          },
        }),
      );

      await Promise.all([
        session.start(),
        (async () => {
          const controller = new AbortController();
          const { signal } = controller;
          const waiter = session.wait(0);

          await assertRejects(
            () => deadline(waiter, 100, { signal }),
            Error,
            "Timeout",
          );
          guard.resolve();
          assertEquals(
            await deadline(waiter, 100, { signal }),
            "Success",
          );

          controller.abort();
        })(),
      ]);
    },
  );

  await t.step(
    "rejects a promise returned by 'wait()' method when an error response is received",
    async () => {
      const guard = deferred<void>();
      const output: unknown[] = [];
      const session = new Session(
        new ReadableStream({
          async start(controller) {
            await guard;
            controller.enqueue(
              encode(buildResponseMessage(0, ["TypeError", "Failed"], null)),
            );
            controller.close();
          },
        }),
        new WritableStream({
          write(chunk) {
            output.push(decode(chunk));
          },
        }),
      );

      await Promise.all([
        session.start(),
        (async () => {
          const controller = new AbortController();
          const { signal } = controller;
          const waiter = session.wait(0);

          await assertRejects(
            () => deadline(waiter, 100, { signal }),
            Error,
            "Timeout",
          );
          guard.resolve();
          await assertRejects(
            () => deadline(waiter, 100, { signal }),
            Error,
            "Failed",
          );

          controller.abort();
        })(),
      ]);
    },
  );

  await t.step(
    "sends arbitrary message when 'send()' method is called",
    async () => {
      const output: unknown[] = [];
      const session = new Session(
        new ReadableStream(),
        new WritableStream({
          write(chunk) {
            output.push(decode(chunk));
          },
        }),
      );

      session.start();

      await session.send(buildRequestMessage(0, "sum", [1, 2]));
      await session.send(buildResponseMessage(0, null, 3));
      await session.send(buildNotificationMessage("sum", [1, 2]));
      assertEquals(
        output,
        [
          buildRequestMessage(0, "sum", [1, 2]),
          buildResponseMessage(0, null, 3),
          buildNotificationMessage("sum", [1, 2]),
        ],
      );
    },
  );
});

async function deadline<T>(
  p: Promise<T>,
  ms: number,
  { signal }: { signal?: AbortSignal } = {},
): Promise<T> {
  const waiter = delay(ms, { signal });
  return await Promise.race([
    waiter.then(() => Promise.reject(new Error("Timeout"))),
    p,
  ]);
}
