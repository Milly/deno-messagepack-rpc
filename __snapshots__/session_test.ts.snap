export const snapshot = {};

snapshot[`Session > invokes a corresponding method and sends back a result when a request is received 1`] = `
[
  [
    1,
    0,
    null,
    3,
  ],
]
`;

snapshot[`Session > invokes a corresponding method and sends back an error result when a request is received (error) 1`] = `
[
  [
    1,
    0,
    [
      "Error",
      "Failed",
    ],
    null,
  ],
]
`;

snapshot[`Session > sends back an error response when a request is received but no corresponding method exists 1`] = `
[
  [
    1,
    0,
    [
      "Error",
      "No MessagePack-RPC method 'sum' exists",
    ],
    null,
  ],
]
`;

snapshot[`Session > invokes onUnexpectedMessage and ignore when unknown message is received 1`] = `
[
  [
    1,
    0,
    null,
    3,
  ],
]
`;
